### PR TITLE
Consolidate Network Bandwidth Metrics

### PR DESCRIPTION
**Pull request by sethalves**
Bandwidth stats are collected and grouped by remote Node type.  Collection is done in LimitedNodeList rather than in various other places.

### CODE REVIEW

File: interface/src/ui/BandwidthDialog.cpp [hunk 1]
```diff
@@ -64,30 +76,44 @@ BandwidthDialog::BandwidthDialog(QWidget* parent, BandwidthRecorder* model) :
...
 
 
 BandwidthDialog::~BandwidthDialog() {
-    delete audioChannelDisplay;
-    delete avatarsChannelDisplay;
-    delete octreeChannelDisplay;
-    delete metavoxelsChannelDisplay;
-    delete totalChannelDisplay;
+    for (unsigned int i=0; i<_CHANNELCOUNT; i++)
+        delete _allChannelDisplays[i];
```
Comment (birarda) [inline interface/src/ui/BandwidthDialog.cpp]: wrap the for loop in curly braces and add spacing on the `=` and `<` operators

File: interface/src/ui/BandwidthDialog.cpp [hunk 2]
```diff
@@ -64,30 +76,44 @@ BandwidthDialog::BandwidthDialog(QWidget* parent, BandwidthRecorder* model) :
...
 
 
-void BandwidthDialog::paintEvent(QPaintEvent* event) {
-    audioChannelDisplay->setLabelText();
-    avatarsChannelDisplay->setLabelText();
-    octreeChannelDisplay->setLabelText();
-    metavoxelsChannelDisplay->setLabelText();
-    totalChannelDisplay->setLabelText();
+void BandwidthDialog::updateTimerTimeout() {
+    for (unsigned int i=0; i<_CHANNELCOUNT; i++)
```
Comment (birarda) [inline interface/src/ui/BandwidthDialog.cpp]: spacing on `=` and `<`

File: interface/src/ui/BandwidthDialog.cpp [hunk 3]
```diff
@@ -64,30 +76,44 @@ BandwidthDialog::BandwidthDialog(QWidget* parent, BandwidthRecorder* model) :
...
+        _allChannelDisplays[i]->bandwidthAverageUpdated();
+}
+
 
+void BandwidthDialog::paintEvent(QPaintEvent* event) {
+    for (unsigned int i=0; i<_CHANNELCOUNT; i++)
+        _allChannelDisplays[i]->Paint();
```
Comment (birarda) [inline interface/src/ui/BandwidthDialog.cpp]: wrap in curly braces, spacing on `=` and `<`

File: interface/src/ui/BandwidthDialog.h [hunk 4]
```diff
@@ -15,36 +15,60 @@
...
+
+
+class BandwidthChannelDisplay : public QObject {
+    Q_OBJECT
+
+ public:
+    BandwidthChannelDisplay(QVector<NodeType_t> nodeTypesToFollow,
+                            QFormLayout* form,
+                            char const* const caption, char const* unitCaption, float unitScale, unsigned colorRGBA);
+    void Paint();
```
Comment (birarda) [inline interface/src/ui/BandwidthDialog.h]: method should be lowercase

File: interface/src/ui/BandwidthDialog.h [hunk 5]
```diff
@@ -15,36 +15,60 @@
...
+
+ private:
+    QVector<NodeType_t> nodeTypesToFollow;
+    QLabel* label;
+    QString strBuf;
+    char const* const caption;
+    char const* unitCaption;
+    float const unitScale;
+    unsigned colorRGBA;
```
Comment (birarda) [inline interface/src/ui/BandwidthDialog.h]: private member variables should be prefixed with underscore

File: interface/src/ui/BandwidthDialog.h [hunk 6]
```diff
@@ -53,15 +77,17 @@ class BandwidthDialog : public QDialog {
...
 
 protected:
-    void paintEvent(QPaintEvent*);
 
     // Emits a 'closed' signal when this dialog is closed.
     void closeEvent(QCloseEvent*);
 
 private:
-    BandwidthRecorder* _model;
+    QTimer *averageUpdateTimer = new QTimer(this);
```
Comment (birarda) [inline interface/src/ui/BandwidthDialog.h]: asterisk should be stuck to `QTimer` here

File: libraries/networking/src/BandwidthRecorder.cpp [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+float BandwidthRecorder::getCachedTotalAverageOutputKilobitsPerSecond() {
+    static qint64 lastCalculated = 0;
+    static float cachedValue = 0.;
+    qint64 now = QDateTime::currentMSecsSinceEpoch();
+    if (now - lastCalculated > 1000.) {
+        lastCalculated = now;
+        cachedValue = getTotalAverageOutputKilobitsPerSecond();
+    }
+    return cachedValue;
+}
```
Comment (birarda) [inline libraries/networking/src/BandwidthRecorder.cpp]: couple of general comments for this file
wrap the if statements and for statements in curly braces
the floating point numbers should be `.0f` terminated

File: libraries/networking/src/LimitedNodeList.cpp [hunk 8]
```diff
@@ -208,6 +206,22 @@ bool LimitedNodeList::packetVersionAndHashMatch(const QByteArray& packet) {
...
+// qint64 LimitedNodeList::readDatagram(char* data, qint64 maxSize, QHostAddress* address = 0, quint16 * port = 0) {
+
+qint64 LimitedNodeList::readDatagram(QByteArray& incomingPacket, QHostAddress* address = 0, quint16 * port = 0) {
+    qint64 result = getNodeSocket().readDatagram(incomingPacket.data(), incomingPacket.size(), address, port);
+
+    SharedNodePointer sendingNode = sendingNodeForPacket(incomingPacket);
+    if (sendingNode) {
+        emit dataReceived(sendingNode->getType(), incomingPacket.size());
+    }
+    else {
```
Comment (birarda) [inline libraries/networking/src/LimitedNodeList.cpp]: this else should be inline on 217

File: interface/src/ui/BandwidthDialog.cpp [hunk 9]
```diff
@@ -64,30 +76,44 @@ BandwidthDialog::BandwidthDialog(QWidget* parent, BandwidthRecorder* model) :
...
-    metavoxelsChannelDisplay = new ChannelDisplay(_model->metavoxelsChannel, form);
-    totalChannelDisplay = new ChannelDisplay(_model->totalChannel, form);
+    QSharedPointer<BandwidthRecorder> bandwidthRecorder = DependencyManager::get<BandwidthRecorder>();
+
+    _allChannelDisplays[0] = _audioChannelDisplay =
+        new BandwidthChannelDisplay({NodeType::AudioMixer}, form, "Audio", "Kbps", 1.0, COLOR0);
+    _allChannelDisplays[1] = _avatarsChannelDisplay =
+        new BandwidthChannelDisplay({NodeType::Agent, NodeType::AvatarMixer}, form, "Avatars", "Kbps", 1.0, COLOR1);
+    _allChannelDisplays[2] = _octreeChannelDisplay =
+        new BandwidthChannelDisplay({NodeType::DomainServer, NodeType::EntityServer}, form, "Octree", "Kbps", 1.0, COLOR2);
```
Comment (birarda) [inline interface/src/ui/BandwidthDialog.cpp]: The Domain server shouldn't be looped into Octree data here, should be its own category

**Comment by birarda** [comment]
Functionality looks great here - made some coding standard comments and one tweak to categories.

My guess is we will later get a point where we want per-node bandwidth stats (when we have multiple types of each server in a domain) - at that time we want want to associate bytes recv/sent with the node and then have the display class group by node type. I think for now in 99% of cases we are running one of each type of server so the current categories should be sufficient.

**Comment by murillodigital** [comment]
add to whitelist

**Comment by hifi-gustavo** [comment]
Build console output for commit: 38162380ade132b9bf1a382dcbae6db1f48e4b1f available at: https://gist.github.com/7769c70a94481efeacad

**Comment by murillodigital** [comment]
@sethalves that fail is not your doing - caused by some recent changes I did to mac slave with multiple versions of Qt on it --- on it

**Comment by murillodigital** [comment]
test this please

**Comment by murillodigital** [comment]
@sethalves (test this please is a trigger for jenkins to build this PR)

**Comment by hifi-gustavo** [comment]
Build console output for commit: 38162380ade132b9bf1a382dcbae6db1f48e4b1f available at: https://gist.github.com/cafb2eca6de3fd76538f

**Comment by hifi-gustavo** [comment]
Build console output for commit: 38162380ade132b9bf1a382dcbae6db1f48e4b1f available at: https://gist.github.com/073a569d6421b06588b3

**Comment by PhilipRosedale** [comment]
I'm going to pull test, as Birarda is sick...

**Comment by birarda** [comment]
@sethalves Philip quickly merged this before the last of the coding standard stuff was covered - can you open up another pull with those changes?

**Comment by birarda** [comment]
Couple of other things I noticed - The bandwidth and packet stats don't have a background behind them now.

screen shot 2015-02-03 at 10 19 35 am

There are also Q_ASSERTS in BandwidthRecorder that we don't need - they can't ever not be true since the `channelType` arg gets casted to `quint8` anyways.

**Pull request merged**